File: python/src/revolverelate/compile/compiler.py
"""Deterministic RelOp → dialect SQL. No LLM. Shared algorithm with typescript/src/compiler.ts."""

from __future__ import annotations

from dataclasses import dataclass, field

from revolverelate.catalog import Engine, get_engine, placeholder, quote_ident


@dataclass
class _Ctx:
    engine: Engine
    params: list = field(default_factory=list)

    def q(self, name: str) -> str:
        return quote_ident(self.engine, name)

    def ph(self) -> str:
        self.params.append(None)  # replaced by emit_param
        return placeholder(self.engine, len(self.params))

    def emit_param(self, value) -> str:
        self.params.append(value)
        return placeholder(self.engine, len(self.params))


def compile_ir(doc: dict, engine: str | Engine) -> tuple[str, list]:
    eng = engine if isinstance(engine, Engine) else get_engine(engine)
    ctx = _Ctx(eng)
    kind = (doc or {}).get("kind") or "query"
    if kind == "query":
        sql = _compile_query(ctx, doc["op"])
    elif kind == "mutate":
        sql = _compile_mutate(ctx, doc["op"])
    elif kind == "txn":
        parts = ["BEGIN"]
        for stmt in doc.get("statements") or []:
            inner, params = compile_ir(stmt, eng)
            parts.append(inner)
            ctx.params.extend(params)
        parts.append("COMMIT")
        sql = "; ".join(parts)
    elif kind == "procedure":
        body = {"kind": "txn", "statements": doc.get("statements") or [{"kind": "mutate", "op": doc["op"]}]}
        return compile_ir(body, eng)
    else:
        raise ValueError(f"Unknown IR kind {kind!r}")
    return sql, list(ctx.params)


def _compile_query(ctx: _Ctx, op: dict) -> str:
    if op.get("op") == "setop":
        left = _compile_query(ctx, op["left"])
        right = _compile_query(ctx, op["right"])
        kind = (op.get("set") or "union").upper()
        if kind == "UNION" and not op.get("all") and ctx.engine.emit_family == "bigquery":
            return f"{left} UNION DISTINCT {right}"
        if op.get("all") and kind == "UNION":
            return f"{left} UNION ALL {right}"
        return f"{left} {kind} {right}"
    if op.get("op") == "with":
        ctes = []
        for cte in op.get("ctes") or []:
            inner = _compile_query(ctx, cte["input"])
            ctes.append(f"{ctx.q(cte['name'])} AS ({inner})")
        body = _compile_query(ctx, op["input"])
        return f"WITH {', '.join(ctes)} {body}"
    sel = _linearize(ctx, op)
    return _render_select(ctx, sel)


@dataclass
class _Select:
    distinct: bool = False
    items: list[str] | None = None
    frm: str = ""
    joins: list[str] = field(default_factory=list)
    where: str | None = None
    groups: list[str] = field(default_factory=list)
    having: str | None = None
    order: list[str] = field(default_factory=list)
    limit: int | None = None
    offset: int | None = None


def _linearize(ctx: _Ctx, op: dict) -> _Select:
    kind = op.get("op")
    if kind == "scan":
        alias = op.get("alias") or op["entity"]
        return _Select(frm=f"{ctx.q(op['entity'])} AS {ctx.q(alias)}")
    if kind == "values":
        cols = op.get("columns") or []
        alias = op.get("alias") or "v"
        raw_rows = op.get("rows") or []
        if cols:
            selects = []
            for row in raw_rows:
                parts = []
                for i, c in enumerate(cols):
                    val = row[i] if i < len(row) else None
                    parts.append(f"{ctx.emit_param(val)} AS {ctx.q(c)}")
                selects.append("SELECT " + ", ".join(parts))
            inner = " UNION ALL ".join(selects) if selects else "SELECT NULL"
            return _Select(frm=f"({inner}) AS {ctx.q(alias)}")
        rows = []
        for row in raw_rows:
            cells = ", ".join(ctx.emit_param(v) for v in row)
            rows.append(f"({cells})")
        values_sql = "VALUES " + ", ".join(rows)
        return _Select(frm=f"({values_sql}) AS {ctx.q(alias)}")
    if kind == "project":
        items = [_project_item(ctx, item) for item in op.get("items") or []]
        sel = _linearize(ctx, op["input"])
        if sel.groups:
            inner = _render_select(ctx, sel)
            sel = _Select(frm=f"({inner}) AS {ctx.q('q')}")
        sel.items = items
        return sel
    if kind == "filter":
        sel = _linearize(ctx, op["input"])
        pred = _compile_expr(ctx, op["predicate"])
        if sel.groups:
            sel.having = f"{sel.having} AND {pred}" if sel.having else pred
        else:
            items_sql = " ".join(sel.items or [])
            if " OVER " in items_sql:
                inner = _render_select(ctx, sel)
                sel = _Select(frm=f"({inner}) AS {ctx.q('q')}")
            sel.where = f"{sel.where} AND {pred}" if sel.where else pred
        return sel
    if kind == "having":
        sel = _linearize(ctx, op["input"])
        pred = _compile_expr(ctx, op["predicate"])
        sel.having = f"{sel.having} AND {pred}" if sel.having else pred
        return sel
    if kind == "join":
        left = _linearize(ctx, op["left"])
        right = _linearize(ctx, op["right"])
        jt = (op.get("joinType") or "inner").upper()
        if jt == "INNER":
            jt = "INNER JOIN"
        elif jt == "LEFT":
            jt = "LEFT JOIN"
        elif jt == "RIGHT":
            jt = "RIGHT JOIN"
        elif jt == "FULL":
            jt = "FULL OUTER JOIN"
        elif jt == "SEMI":
            jt = "INNER JOIN"
        elif jt == "ANTI":
            jt = "LEFT JOIN"
        else:
            jt = f"{jt} JOIN"
        on = " AND ".join(_compile_expr(ctx, p) for p in op.get("on") or [])
        right_from = right.frm
        left.joins.append(f"{jt} {right_from}" + (f" ON {on}" if on else ""))
        left.joins.extend(right.joins)
        if right.where:
            left.where = f"{left.where} AND {right.where}" if left.where else right.where
        return left
    if kind == "aggregate":
        groups = [_compile_expr(ctx, g) for g in op.get("groups") or []]
        aggs = [_project_item(ctx, a) for a in op.get("aggs") or []]
        sel = _linearize(ctx, op["input"])
        sel.groups = groups
        sel.items = groups + aggs
        return sel
    if kind == "sort":
        sel = _linearize(ctx, op["input"])
        keys = []
        for key in op.get("keys") or []:
            expr = _compile_expr(ctx, key["expr"])
            direction = (key.get("direction") or "ASC").upper()
            keys.append(f"{expr} {direction}")
        sel.order = keys
        return sel
    if kind == "limit":
        sel = _linearize(ctx, op["input"])
        sel.limit = int(op["count"])
        if op.get("offset"):
            sel.offset = int(op["offset"])
        return sel
    if kind == "distinct":
        sel = _linearize(ctx, op["input"])
        sel.distinct = True
        return sel
    if kind == "window":
        sel = _linearize(ctx, op["input"])
        extra = [_project_item(ctx, item) for item in op.get("items") or []]
        sel.items = (sel.items or ["*"]) + extra
        return sel
    # wrap unknown as subquery
    inner = _compile_query(ctx, op) if kind in {"setop", "with"} else _render_select(ctx, _Select())
    return _Select(frm=f"({inner}) AS {ctx.q('q')}")


def _project_item(ctx: _Ctx, item: dict) -> str:
    expr = item["expr"] if "expr" in item and isinstance(item.get("expr"), dict) else item
    sql = _compile_expr(ctx, expr)
    alias = item.get("alias")
    if alias:
        return f"{sql} AS {ctx.q(alias)}"
    return sql


def _compile_expr(ctx: _Ctx, expr: dict | None) -> str:
    if not expr:
        return "1"
    kind = expr.get("expr")
    if kind == "star":
        entity = expr.get("entity") or expr.get("alias")
        return f"{ctx.q(entity)}.*" if entity else "*"
    if kind == "col":
        entity = expr.get("entity") or expr.get("alias")
        attr = expr.get("attr")
        if entity and attr:
            return f"{ctx.q(entity)}.{ctx.q(attr)}"
        if attr:
            return ctx.q(attr)
        return ctx.q(entity or "col")
    if kind == "lit":
        return ctx.emit_param(expr.get("value"))
    if kind == "param":
        return ctx.emit_param(expr.get("value"))
    if kind == "agg":
        fn = (expr.get("fn") or "count").upper()
        inner = expr.get("input") or {"expr": "star"}
        if inner.get("expr") == "star" and fn == "COUNT" and not expr.get("distinct"):
            return f"{fn}(*)"
        if expr.get("distinct"):
            return f"{fn}(DISTINCT {_compile_expr(ctx, inner)})"
        return f"{fn}({_compile_expr(ctx, inner)})"
    if kind == "bin":
        op = expr.get("op") or "="
        if op.lower() == "and":
            return f"({_compile_expr(ctx, expr.get('left'))} AND {_compile_expr(ctx, expr.get('right'))})"
        if op.lower() == "or":
            return f"({_compile_expr(ctx, expr.get('left'))} OR {_compile_expr(ctx, expr.get('right'))})"
        if op.lower() == "in":
            values = expr.get("right", {}).get("value")
            if isinstance(values, (list, tuple)):
                ph = ", ".join(ctx.emit_param(v) for v in values)
                return f"{_compile_expr(ctx, expr['left'])} IN ({ph})"
        if op.lower() == "like":
            return f"{_compile_expr(ctx, expr['left'])} LIKE {_compile_expr(ctx, expr['right'])}"
        if op.lower() == "not like":
            return f"{_compile_expr(ctx, expr['left'])} NOT LIKE {_compile_expr(ctx, expr['right'])}"
        if op.lower() == "not in":
            values = expr.get("right", {}).get("value")
            if isinstance(values, (list, tuple)):
                ph = ", ".join(ctx.emit_param(v) for v in values)
                return f"{_compile_expr(ctx, expr['left'])} NOT IN ({ph})"
        if op.lower() == "between":
            values = expr.get("right", {}).get("value")
            if isinstance(values, (list, tuple)) and len(values) >= 2:
                return (
                    f"{_compile_expr(ctx, expr['left'])} BETWEEN {ctx.emit_param(values[0])} "
                    f"AND {ctx.emit_param(values[1])}"
                )
        if op.lower() == "is null":
            return f"{_compile_expr(ctx, expr.get('left'))} IS NULL"
        if op.lower() == "is not null":
            return f"{_compile_expr(ctx, expr.get('left'))} IS NOT NULL"
        return f"{_compile_expr(ctx, expr.get('left'))} {op} {_compile_expr(ctx, expr.get('right'))}"
    if kind == "over":
        fn = (expr.get("fn") or "rank").upper()
        inner = expr.get("input")
        call = f"{fn}()" if inner is None and fn in {"RANK", "DENSE_RANK", "ROW_NUMBER"} else f"{fn}({_compile_expr(ctx, inner) if inner else '*'})"
        parts = []
        parts_p = [_compile_expr(ctx, p) for p in expr.get("partition") or []]
        if parts_p:
            parts.append("PARTITION BY " + ", ".join(parts_p))
        parts_o = []
        for key in expr.get("order") or []:
            direction = (key.get("direction") or "ASC").upper()
            parts_o.append(f"{_compile_expr(ctx, key['expr'])} {direction}")
        if parts_o:
            parts.append("ORDER BY " + ", ".join(parts_o))
        return f"{call} OVER ({' '.join(parts)})"
    if kind == "un":
        op = (expr.get("op") or "not").upper()
        return f"{op} ({_compile_expr(ctx, expr.get('input'))})"
    if kind == "fn":
        fn = (expr.get("fn") or "FN").upper()
        args = ", ".join(_compile_expr(ctx, a) for a in expr.get("args") or [])
        return f"{fn}({args})"
    if kind == "case":
        parts = ["CASE"]
        for arm in expr.get("whens") or []:
            parts.append(f"WHEN {_compile_expr(ctx, arm.get('when'))} THEN {_compile_expr(ctx, arm.get('then'))}")
        if expr.get("else") is not None:
            parts.append(f"ELSE {_compile_expr(ctx, expr.get('else'))}")
        parts.append("END")
        return " ".join(parts)
    return "1"


def _render_select(ctx: _Ctx, sel: _Select) -> str:
    items = ", ".join(sel.items) if sel.items else "*"
    distinct = "DISTINCT " if sel.distinct else ""
    top = ""
    limit_sql = ""
    if sel.limit is not None:
        if ctx.engine.limit_style == "top":
            top = f"TOP {sel.limit} "
        elif ctx.engine.limit_style == "fetch":
            limit_sql = f" FETCH FIRST {sel.limit} ROWS ONLY"
        else:
            limit_sql = f" LIMIT {sel.limit}"
            if sel.offset:
                limit_sql += f" OFFSET {sel.offset}"
    sql = f"SELECT {distinct}{top}{items} FROM {sel.frm}"
    if sel.joins:
        sql += " " + " ".join(sel.joins)
    if sel.where:
        sql += f" WHERE {sel.where}"
    if sel.groups:
        sql += " GROUP BY " + ", ".join(sel.groups)
    if sel.having:
        sql += f" HAVING {sel.having}"
    if sel.order:
        sql += " ORDER BY " + ", ".join(sel.order)
    sql += limit_sql
    return sql


def _compile_mutate(ctx: _Ctx, op: dict) -> str:
    kind = op.get("op")
    if kind == "insert":
        cols = ", ".join(ctx.q(c) for c in op.get("columns") or [])
        if op.get("input"):
            select = _compile_query(ctx, op["input"])
            return f"INSERT INTO {ctx.q(op['entity'])} ({cols}) {select}"
        rows_sql = []
        for row in op.get("rows") or []:
            rows_sql.append("(" + ", ".join(ctx.emit_param(v) for v in row) + ")")
        return f"INSERT INTO {ctx.q(op['entity'])} ({cols}) VALUES {', '.join(rows_sql)}"
    if kind == "update":
        sets = []
        for col, expr in (op.get("set") or {}).items():
            if isinstance(expr, dict):
                sets.append(f"{ctx.q(col)} = {_compile_expr(ctx, expr)}")
            else:
                sets.append(f"{ctx.q(col)} = {ctx.emit_param(expr)}")
        sql = f"UPDATE {ctx.q(op['entity'])} SET {', '.join(sets)}"
        if op.get("predicate"):
            sql += f" WHERE {_compile_expr(ctx, op['predicate'])}"
        return sql
    if kind == "delete":
        sql = f"DELETE FROM {ctx.q(op['entity'])}"
        if op.get("predicate"):
            sql += f" WHERE {_compile_expr(ctx, op['predicate'])}"
        return sql
    if kind == "merge":
        target = ctx.q(op["target"])
        source = _compile_query(ctx, op["source"])
        on = _compile_expr(ctx, op["on"])
        if ctx.engine.emit_family in {"postgres", "sqlite", "duckdb"}:
            # compile as INSERT ... SELECT with ON CONFLICT when keys provided
            cols = op.get("columns") or []
            col_sql = ", ".join(ctx.q(c) for c in cols)
            return f"INSERT INTO {target} ({col_sql}) {source}"
        return f"MERGE INTO {target} USING ({source}) AS {ctx.q('src')} ON {on}"
    if kind == "call":
        args = ", ".join(ctx.emit_param(a) for a in op.get("args") or [])
        return f"CALL {ctx.q(op['procedure'])}({args})"
    raise ValueError(f"Unknown mutate op {kind!r}")

File: typescript/src/compiler.ts
/** Deterministic RelOp → dialect SQL. Shared algorithm with python compile/compiler.py. */

import { type Engine, getEngine, placeholder, quoteIdent } from "./catalog.ts";

type Expr = Record<string, any>;
type Op = Record<string, any>;
type Doc = { kind?: string; op?: Op; statements?: Doc[]; name?: string };

type Ctx = { engine: Engine; params: unknown[] };

function q(ctx: Ctx, name: string): string {
  return quoteIdent(ctx.engine, name);
}

function emitParam(ctx: Ctx, value: unknown): string {
  ctx.params.push(value);
  return placeholder(ctx.engine, ctx.params.length);
}

export function compileIr(doc: Doc, engine: string | Engine): { sql: string; params: unknown[] } {
  const eng = typeof engine === "string" ? getEngine(engine) : engine;
  const ctx: Ctx = { engine: eng, params: [] };
  const kind = doc.kind || "query";
  let sql = "";
  if (kind === "query") {
    sql = compileQuery(ctx, doc.op!);
  } else if (kind === "mutate") {
    sql = compileMutate(ctx, doc.op!);
  } else if (kind === "txn") {
    const parts = ["BEGIN"];
    for (const stmt of doc.statements || []) {
      const inner = compileIr(stmt, eng);
      parts.push(inner.sql);
      ctx.params.push(...inner.params);
    }
    parts.push("COMMIT");
    sql = parts.join("; ");
  } else if (kind === "procedure") {
    return compileIr(
      { kind: "txn", statements: doc.statements || [{ kind: "mutate", op: doc.op }] },
      eng,
    );
  } else {
    throw new Error(`Unknown IR kind ${kind}`);
  }
  return { sql, params: ctx.params };
}

function compileQuery(ctx: Ctx, op: Op): string {
  if (op.op === "setop") {
    const left = compileQuery(ctx, op.left);
    const right = compileQuery(ctx, op.right);
    const kind = String(op.set || "union").toUpperCase();
    if (kind === "UNION" && !op.all && ctx.engine.emitFamily === "bigquery") {
      return `${left} UNION DISTINCT ${right}`;
    }
    if (op.all && kind === "UNION") return `${left} UNION ALL ${right}`;
    return `${left} ${kind} ${right}`;
  }
  if (op.op === "with") {
    const ctes = (op.ctes || []).map(
      (cte: Op) => `${q(ctx, cte.name)} AS (${compileQuery(ctx, cte.input)})`,
    );
    return `WITH ${ctes.join(", ")} ${compileQuery(ctx, op.input)}`;
  }
  return renderSelect(ctx, linearize(ctx, op));
}

type Select = {
  distinct: boolean;
  items: string[] | null;
  frm: string;
  joins: string[];
  where: string | null;
  groups: string[];
  having: string | null;
  order: string[];
  limit: number | null;
  offset: number | null;
};

function emptySelect(): Select {
  return {
    distinct: false,
    items: null,
    frm: "",
    joins: [],
    where: null,
    groups: [],
    having: null,
    order: [],
    limit: null,
    offset: null,
  };
}

function linearize(ctx: Ctx, op: Op): Select {
  const kind = op.op;
  if (kind === "scan") {
    const alias = op.alias || op.entity;
    const sel = emptySelect();
    sel.frm = `${q(ctx, op.entity)} AS ${q(ctx, alias)}`;
    return sel;
  }
  if (kind === "values") {
    const alias = op.alias || "v";
    const cols = op.columns || [];
    const rawRows: unknown[][] = op.rows || [];
    const sel = emptySelect();
    if (cols.length) {
      const selects = rawRows.map((row) => {
        const parts = cols.map((c: string, i: number) => `${emitParam(ctx, row[i])} AS ${q(ctx, c)}`);
        return `SELECT ${parts.join(", ")}`;
      });
      const inner = selects.length ? selects.join(" UNION ALL ") : "SELECT NULL";
      sel.frm = `(${inner}) AS ${q(ctx, alias)}`;
      return sel;
    }
    const rows = rawRows.map(
      (row: unknown[]) => `(${row.map((v) => emitParam(ctx, v)).join(", ")})`,
    );
    sel.frm = `(VALUES ${rows.join(", ")}) AS ${q(ctx, alias)}`;
    return sel;
  }
  if (kind === "project") {
    const items = (op.items || []).map((item: Op) => projectItem(ctx, item));
    const sel = linearize(ctx, op.input);
    sel.items = items;
    return sel;
  }
  if (kind === "filter") {
    const sel = linearize(ctx, op.input);
    const pred = compileExpr(ctx, op.predicate);
    if (sel.groups.length) sel.having = sel.having ? `${sel.having} AND ${pred}` : pred;
    else {
      const itemsSql = (sel.items || []).join(" ");
      if (itemsSql.includes(" OVER ")) {
        const inner = renderSelect(ctx, sel);
        const wrapped = emptySelect();
        wrapped.frm = `(${inner}) AS ${q(ctx, "q")}`;
        wrapped.where = pred;
        return wrapped;
      }
      sel.where = sel.where ? `${sel.where} AND ${pred}` : pred;
    }
    return sel;
  }
  if (kind === "having") {
    const sel = linearize(ctx, op.input);
    const pred = compileExpr(ctx, op.predicate);
    sel.having = sel.having ? `${sel.having} AND ${pred}` : pred;
    return sel;
  }
  if (kind === "join") {
    const left = linearize(ctx, op.left);
    const right = linearize(ctx, op.right);
    let jt = String(op.joinType || "inner").toUpperCase();
    if (jt === "INNER") jt = "INNER JOIN";
    else if (jt === "LEFT") jt = "LEFT JOIN";
    else if (jt === "RIGHT") jt = "RIGHT JOIN";
    else if (jt === "FULL") jt = "FULL OUTER JOIN";
    else if (jt === "SEMI") jt = "INNER JOIN";
    else if (jt === "ANTI") jt = "LEFT JOIN";
    else jt = `${jt} JOIN`;
    const on = (op.on || []).map((p: Expr) => compileExpr(ctx, p)).join(" AND ");
    left.joins.push(`${jt} ${right.frm}` + (on ? ` ON ${on}` : ""));
    left.joins.push(...right.joins);
    if (right.where) left.where = left.where ? `${left.where} AND ${right.where}` : right.where;
    return left;
  }
  if (kind === "aggregate") {
    const groups = (op.groups || []).map((g: Expr) => compileExpr(ctx, g));
    const aggs = (op.aggs || []).map((a: Op) => projectItem(ctx, a));
    const sel = linearize(ctx, op.input);
    sel.groups = groups;
    sel.items = groups.concat(aggs);
    return sel;
  }
  if (kind === "sort") {
    const sel = linearize(ctx, op.input);
    sel.order = (op.keys || []).map((key: Op) => {
      const expr = compileExpr(ctx, key.expr);
      const direction = String(key.direction || "ASC").toUpperCase();
      return `${expr} ${direction}`;
    });
    return sel;
  }
  if (kind === "limit") {
    const sel = linearize(ctx, op.input);
    sel.limit = Number(op.count);
    if (op.offset) sel.offset = Number(op.offset);
    return sel;
  }
  if (kind === "distinct") {
    const sel = linearize(ctx, op.input);
    sel.distinct = true;
    return sel;
  }
  if (kind === "window") {
    const sel = linearize(ctx, op.input);
    const extra = (op.items || []).map((item: Op) => projectItem(ctx, item));
    sel.items = (sel.items || ["*"]).concat(extra);
    return sel;
  }
  const inner = compileQuery(ctx, op);
  const sel = emptySelect();
  sel.frm = `(${inner}) AS ${q(ctx, "q")}`;
  return sel;
}

function projectItem(ctx: Ctx, item: Op): string {
  const expr = item.expr && typeof item.expr === "object" ? item.expr : item;
  const sql = compileExpr(ctx, expr);
  if (item.alias) return `${sql} AS ${q(ctx, item.alias)}`;
  return sql;
}

function compileExpr(ctx: Ctx, expr: Expr | undefined): string {
  if (!expr) return "1";
  const kind = expr.expr;
  if (kind === "star") {
    const entity = expr.entity || expr.alias;
    return entity ? `${q(ctx, entity)}.*` : "*";
  }
  if (kind === "col") {
    const entity = expr.entity || expr.alias;
    const attr = expr.attr;
    if (entity && attr) return `${q(ctx, entity)}.${q(ctx, attr)}`;
    if (attr) return q(ctx, attr);
    return q(ctx, entity || "col");
  }
  if (kind === "lit" || kind === "param") return emitParam(ctx, expr.value);
  if (kind === "agg") {
    const fn = String(expr.fn || "count").toUpperCase();
    const inner = expr.input || { expr: "star" };
    if (inner.expr === "star" && fn === "COUNT") return `${fn}(*)`;
    return `${fn}(${compileExpr(ctx, inner)})`;
  }
  if (kind === "bin") {
    const op = expr.op || "=";
    if (String(op).toLowerCase() === "and") {
      return `(${compileExpr(ctx, expr.left)} AND ${compileExpr(ctx, expr.right)})`;
    }
    if (String(op).toLowerCase() === "or") {
      return `(${compileExpr(ctx, expr.left)} OR ${compileExpr(ctx, expr.right)})`;
    }
    if (String(op).toLowerCase() === "in") {
      const values = expr.right?.value;
      if (Array.isArray(values)) {
        const ph = values.map((v: unknown) => emitParam(ctx, v)).join(", ");
        return `${compileExpr(ctx, expr.left)} IN (${ph})`;
      }
    }
    if (String(op).toLowerCase() === "like") {
      return `${compileExpr(ctx, expr.left)} LIKE ${compileExpr(ctx, expr.right)}`;
    }
    return `${compileExpr(ctx, expr.left)} ${op} ${compileExpr(ctx, expr.right)}`;
  }
  if (kind === "over") {
    const fn = String(expr.fn || "rank").toUpperCase();
    const inner = expr.input;
    const call =
      inner == null && ["RANK", "DENSE_RANK", "ROW_NUMBER"].includes(fn)
        ? `${fn}()`
        : `${fn}(${inner ? compileExpr(ctx, inner) : "*"})`;
    const parts: string[] = [];
    const partsP = (expr.partition || []).map((p: Expr) => compileExpr(ctx, p));
    if (partsP.length) parts.push("PARTITION BY " + partsP.join(", "));
    const partsO = (expr.order || []).map((key: Op) => {
      const direction = String(key.direction || "ASC").toUpperCase();
      return `${compileExpr(ctx, key.expr)} ${direction}`;
    });
    if (partsO.length) parts.push("ORDER BY " + partsO.join(", "));
    return `${call} OVER (${parts.join(" ")})`;
  }
  if (kind === "un") {
    const op = String(expr.op || "not").toUpperCase();
    return `${op} (${compileExpr(ctx, expr.input)})`;
  }
  if (kind === "fn") {
    const fn = String(expr.fn || "FN").toUpperCase();
    const args = (expr.args || []).map((a: Expr) => compileExpr(ctx, a)).join(", ");
    return `${fn}(${args})`;
  }
  if (kind === "case") {
    const parts = ["CASE"];
    for (const arm of expr.whens || []) {
      parts.push(`WHEN ${compileExpr(ctx, arm.when)} THEN ${compileExpr(ctx, arm.then)}`);
    }
    if (expr.else != null) parts.push(`ELSE ${compileExpr(ctx, expr.else)}`);
    parts.push("END");
    return parts.join(" ");
  }
  return "1";
}

function renderSelect(ctx: Ctx, sel: Select): string {
  const items = sel.items ? sel.items.join(", ") : "*";
  const distinct = sel.distinct ? "DISTINCT " : "";
  let top = "";
  let limitSql = "";
  if (sel.limit !== null) {
    if (ctx.engine.limitStyle === "top") top = `TOP ${sel.limit} `;
    else if (ctx.engine.limitStyle === "fetch") limitSql = ` FETCH FIRST ${sel.limit} ROWS ONLY`;
    else {
      limitSql = ` LIMIT ${sel.limit}`;
      if (sel.offset) limitSql += ` OFFSET ${sel.offset}`;
    }
  }
  let sql = `SELECT ${distinct}${top}${items} FROM ${sel.frm}`;
  if (sel.joins.length) sql += " " + sel.joins.join(" ");
  if (sel.where) sql += ` WHERE ${sel.where}`;
  if (sel.groups.length) sql += " GROUP BY " + sel.groups.join(", ");
  if (sel.having) sql += ` HAVING ${sel.having}`;
  if (sel.order.length) sql += " ORDER BY " + sel.order.join(", ");
  return sql + limitSql;
}

function compileMutate(ctx: Ctx, op: Op): string {
  const kind = op.op;
  if (kind === "insert") {
    const cols = (op.columns || []).map((c: string) => q(ctx, c)).join(", ");
    if (op.input) return `INSERT INTO ${q(ctx, op.entity)} (${cols}) ${compileQuery(ctx, op.input)}`;
    const rows = (op.rows || []).map(
      (row: unknown[]) => `(${row.map((v) => emitParam(ctx, v)).join(", ")})`,
    );
    return `INSERT INTO ${q(ctx, op.entity)} (${cols}) VALUES ${rows.join(", ")}`;
  }
  if (kind === "update") {
    const sets = Object.entries(op.set || {}).map(([col, expr]) => {
      if (expr && typeof expr === "object" && (expr as Expr).expr) {
        return `${q(ctx, col)} = ${compileExpr(ctx, expr as Expr)}`;
      }
      return `${q(ctx, col)} = ${emitParam(ctx, expr)}`;
    });
    let sql = `UPDATE ${q(ctx, op.entity)} SET ${sets.join(", ")}`;
    if (op.predicate) sql += ` WHERE ${compileExpr(ctx, op.predicate)}`;
    return sql;
  }
  if (kind === "delete") {
    let sql = `DELETE FROM ${q(ctx, op.entity)}`;
    if (op.predicate) sql += ` WHERE ${compileExpr(ctx, op.predicate)}`;
    return sql;
  }
  if (kind === "merge") {
    const target = q(ctx, op.target);
    const source = compileQuery(ctx, op.source);
    const on = compileExpr(ctx, op.on);
    if (["postgres", "sqlite", "duckdb"].includes(ctx.engine.emitFamily)) {
      const cols = (op.columns || []).map((c: string) => q(ctx, c)).join(", ");
      return `INSERT INTO ${target} (${cols}) ${source}`;
    }
    return `MERGE INTO ${target} USING (${source}) AS ${q(ctx, "src")} ON ${on}`;
  }
  if (kind === "call") {
    const args = (op.args || []).map((a: unknown) => emitParam(ctx, a)).join(", ");
    return `CALL ${q(ctx, op.procedure)}(${args})`;
  }
  throw new Error(`Unknown mutate op ${kind}`);
}
